### PR TITLE
[USM] Add telemetry for interesting frame split on multiple TCP payloads

### PR DESCRIPTION
**Pull request by amitslavin**
### What does this PR do?

Adds telemetry for interesting frame split on multiple TCP payloads

### Motivation

To improve the accuracy of our current HTTP/2 processing, we would like to understand how frequently we encounter scenarios in which an interesting frame is split into multiple TCP payloads. Currently, our agent does not support this behavior.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/ebpf/c/protocols/http2/decoding-defs.h [hunk 1]
```diff
@@ -225,6 +225,7 @@ typedef struct {
 // exceeding_max_interesting_frames		Count of times we reached the max number of frames per iteration.
 // exceeding_max_frames_to_filter		Count of times we have left with more frames to filter than the max number of frames to filter.
 // path_size_bucket                     Count of path sizes and divided into buckets.
+// frames_split_count                   Count of times we tried to read more data than the end of the data end.
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding-defs.h:228]: let's find a better name

File: pkg/network/ebpf/c/protocols/http2/decoding-tls.h [hunk 2]
```diff
@@ -564,6 +564,11 @@ static __always_inline void tls_find_relevant_frames(tls_dispatcher_arguments_t
             iteration_value->frames_array[iteration_value->frames_count].offset = info->data_off;
             iteration_value->frames_count++;
         }
+
+        if (info->data_off + current_frame.length > info->data_end) {
+            __sync_fetch_and_add(&http2_tel->frames_split_count, 1);
+        }
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding-tls.h:570]: let's wrap in a function as the implementation is same everywhere

File: pkg/network/usm/usm_http2_monitor_test.go [hunk 3]
```diff
@@ -499,10 +499,11 @@ func (s *usmHTTP2Suite) TestRawTraffic() {
 	}
 
 	tests := []struct {
-		name              string
-		skip              bool
-		messageBuilder    func() [][]byte
-		expectedEndpoints map[usmhttp.Key]int
+		name                      string
+		skip                      bool
+		validateTCPSplitTelemetry bool
```
Comment (guyarb) [inline pkg/network/usm/usm_http2_monitor_test.go:504]: this flag is used only once for a single subtest
I think it is better to move it out to a different test

File: pkg/network/usm/usm_http2_monitor_test.go [hunk 4]
```diff
@@ -1133,6 +1134,22 @@ func (s *usmHTTP2Suite) TestRawTraffic() {
...
+		{
+			name: "validate message split into 2 tcp segments",
+			// The purpose of this test is to validate that we cannot handle reassembled tcp segments.
+			messageBuilder: func() [][]byte {
+				a := newFramer().
+					writeHeaders(t, 1, usmhttp2.HeadersFrameOptions{Headers: testHeaders()}).
+					writeData(t, 1, true, emptyBody).bytes()
+				return [][]byte{
+					a[:10],
+					a[10:],
```
Comment (guyarb) [inline pkg/network/usm/usm_http2_monitor_test.go:1146]: why 10? document

File: pkg/network/ebpf/c/protocols/http2/decoding-common.h [hunk 5]
```diff
@@ -161,4 +161,11 @@ static __always_inline void reset_frame(http2_frame_t *out) {
     *out = (http2_frame_t){ 0 };
 }
 
+// check_frame_split checks if the frame is split into multiple tcp payloads, if so, it increments the split counter.
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding-common.h:164]:
```suggestion
// checks if the frame is split into multiple tcp payloads, if so, it increments the split counter.
```

Comment (guyarb) [reply pkg/network/ebpf/c/protocols/http2/decoding-common.h:164]: shouldn't you check that only for interesting frame?

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/ebpf/c/protocols/http2/decoding-common.h (hunk 5 above)
Comment (amitslavin) [reply pkg/network/ebpf/c/protocols/http2/decoding-common.h:164]: changed it to be included only in interesting frames

**Review by guyarb — APPROVED** [review]

**Review by guyarb — APPROVED** [review]

**Pull request merged**